src/main/java/org/folio/circulation/domain/CreateRequestService.java

```diff
@@ -155,4 +155,44 @@ private static RequestAndRelatedRecords setRequestQueuePosition(
   private static boolean isTheSameRequester(RequestAndRelatedRecords it, Request that) {
     return Objects.equals(it.getUserId(), that.getUserId());
   }
+
+  private Result<RequestAndRelatedRecords> refuseWhenUserHasAlreadyBeenLoanedItem(Result<RequestAndRelatedRecords> result,
+      LoanRepository loanRepository) {
+
+    RequestAndRelatedRecords requestAndRelatedRecords = result.value();
+
+    if(requestAndRelatedRecords != null) {
+      Request request = requestAndRelatedRecords.getRequest();
+    
+      Loan loan = null;
+      try {
+        loan = loanRepository
```

Comment:
We typically use the methods of `CompletableFuture` that allow for actions to be undertaken after they complete, rather than blocking.

Is this blocking (and catching the future exceptions) because the alternative doesn't work with here, or would be useful for me to propose an alternative way of doing this?

Reply:
I think there are a couple of alternatives:
`
    return result
      .map(RequestAndRelatedRecords::getRequest)
      .combineToResultAfter(loanRepository::findOpenLoanForRequest,
        (request, loan) -> {
          if (loan != null && loan.getUserId().equals(request.getUserId())) {
            Map<String, String> parameters = new HashMap<>();
            parameters.put("userId", request.getUserId());
            parameters.put("loanId", loan.getId());
            String message = "This requester currently has this item on loan.";
            return failedValidation(new ValidationError(message, parameters));
          } else {
            return result;
          }
        });
`

`
  private CompletableFuture<Result<RequestAndRelatedRecords>> refuseWhenUserHasAlreadyBeenLoanedItem(
    Result<RequestAndRelatedRecords> result) {

    return result
      .map(RequestAndRelatedRecords::getRequest)
      .failAfter(this::hasOpenLoanForSameUser,
        this::failureWhenItemAlreadyLoanedToSameUser)
      .thenApply(validationResult -> result.combine(validationResult,
        RequestAndRelatedRecords::withRequest));
  }

  private HttpFailure failureWhenItemAlreadyLoanedToSameUser(Request request) {
    Map<String, String> parameters = new HashMap<>();
    parameters.put("userId", request.getUserId());
    parameters.put("itemId", request.getItemId());
    String message = "This requester currently has this item on loan.";

    return singleValidationError(new ValidationError(message, parameters));
  }

  private CompletableFuture<Result<Boolean>> hasOpenLoanForSameUser(Request request) {
    return loanRepository.findOpenLoanForRequest(request)
      .thenApply(result -> result.map(
        loan -> loan != null && loan.getUserId().equals(request.getUserId())));
  }
`

The first requires a new method on the result class, the second only requires extracting a function for the check.

Which do you prefer?

Reply:
I have consulted with William and we have taken a slightly different approach. I'm eager to get your thoughts.